File: src/main/java/br/campos/poi/model/factory/AbstractDataFactory.java
package br.campos.poi.model.factory;

import br.campos.poi.model.abst.Document;
import br.campos.poi.model.abst.ObjectDTO;
import br.campos.poi.model.document.POIDocument;
import br.campos.poi.model.transfer.POIDTO;

/**
 * Fábrica abstrata de DocumentFactory e ObjectDTOFactory
 * 
 * @author caio
 */
public abstract class AbstractDataFactory {

	/**
	 * Gera a fábrica adequada ao tipo de dado
	 * 
	 * @param dto Objeto de amostra para a identificação da fábrica necessária
	 * @return
	 */
	public static DocumentFactory getFactory(ObjectDTO dto) {
		switch (dto.getType()) {
		case POI_TYPE:
			return POIDocument::new;
		default:
			return null;
		}
	}

	/**
	 * Gera a fábrica adequada ao tipo de dado
	 * 
	 * @param dto Objeto de amostra para a identificação da fábrica necessária
	 * @return
	 */
	public static ObjectDTOFactory getFactory(Document doc) {
		switch (doc.getType()) {
		case POI_TYPE:
			return POIDTO::new;
		default:
			return null;
		}
	}
}

File: src/main/kotlin/br/campos/poi/model/factory/AbstractDataFactory.kt
package br.campos.poi.model.factory

import br.campos.poi.model.abst.Document
import br.campos.poi.model.abst.ObjectDTO
import br.campos.poi.model.document.POIDocument
import br.campos.poi.model.transfer.POIDTO
import br.campos.poi.util.DataType

/**
 * Fábrica abstrata de DocumentFactory e ObjectDTOFactory
 *
 * @author caio
 */
object AbstractDataFactory {

	/**
	 * Gera a fábrica adequada ao tipo de dado
	 *
	 * @param dto Objeto de amostra para a identificação da fábrica necessária
	 */
	fun getFactory(dto: ObjectDTO): DocumentFactory? = when (dto.type) {
		DataType.POI_TYPE -> DocumentFactory { d -> POIDocument(d) }
		else -> null
	}

	/**
	 * Gera a fábrica adequada ao tipo de dado
	 *
	 * @param doc Objeto de amostra para a identificação da fábrica necessária
	 */
	fun getFactory(doc: Document): ObjectDTOFactory? = when (doc.type) {
		DataType.POI_TYPE -> ObjectDTOFactory { d -> POIDTO(d) }
		else -> null
	}
}
